bd-report-convert/src/glue.cpp

```diff
@@ -0,0 +1,102 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+#include <cstdint>
+#include <fstream>
+#include <iostream>
+#include <ostream>
+#include <string>
+
+#include <flatbuffers/idl.h>
+#include <flatbuffers/util.h>
+#include <flatbuffers/minireflect.h>
+
+#include "flatbuffers/report_generated.h"
+
+enum BDReaderErr {
+  BDReaderErrCouldNotReadSpec = -1,
+  BDReaderErrCouldNotReadInputFile = -2,
+  BDReaderErrCouldNotParseSpec = -3,
+  BDReaderErrCouldNotParseInputData = -4,
+  BDReaderErrCouldNotValidateOutput = -5,
+};
+
+#define SPECDIR "api/src/bitdrift_public/fbs/issue-reporting/v1"
+const char *SPECPATH = SPECDIR "/report.fbs";
+
+using namespace bitdrift_public::fbs::issue_reporting::v1;
+extern "C" {
+int bdrr_print_json(char *bin_data_path) {
+  std::ifstream data_file;
+  data_file.open(bin_data_path, std::ios::binary | std::ios::in);
+  data_file.seekg(0,std::ios::end);
+
+  int length = data_file.tellg();
+  data_file.seekg(0,std::ios::beg);
+  char *bin_data = new char[length];
+  data_file.read(bin_data, length);
+  data_file.close();
+
+  auto output = flatbuffers::FlatBufferToString(
+      reinterpret_cast<uint8_t *>(bin_data),
+      ReportTypeTable(),
+      false, true, "", true);
+
+  std::cout << output << std::endl;
+  return 0;
+}
+
+/**
+ * Create a buffer containing the binary representation of a flatbuffer file
+ *
+ * @param json_data_path  the path to the data to convert in JSON format
+ * @param length_or_err   destination for the length of the buffer or an error
+ *
+ * @return an allocated buffer containing the binary data or null if parsing
+ *         failed and the value of length_or_err should be checked for the
+ *         problem
+ */
+uint8_t *bdrr_make_bin_from_json(
+    char *json_data_path,
+    int32_t */* out */length_or_err
+) {
+  std::string json_data;
+  std::string schema_data;
+
+  if (!flatbuffers::LoadFile(SPECPATH, false, &schema_data)) {
```

Comment:
Does this mean that we also need to have this available at runtime? probably OK as long as we're expecting people to run via cargo run but I imagine this would break otherwise? We could embed the file into the binary without too much trouble but not sure if it matters